proposals/1984_ca-bundle-from-secret-ref/README.md

```diff
@@ -0,0 +1,134 @@
+Once approved and implemented, they become historical documents.
+If you are reading an old proposal, please be aware that the
+features described herein might have continued to evolve since.
+
+# CA Bundles from secret refs
+
+## Problem Statement
+
+Many Pinniped custom resources (CRs) have an inline `certificateAuthorityData` to specify a base-64 encoded CA Bundle.
+These fields require manual intervention to set and update.
+Customers who use Kubernetes-native certificate-management tooling (such as `cert-manager`) may wish to keep their
+CA bundles in a secret that their tooling can manage.
+
+## Proposed solution
+
+The users can specify a `secretName` anywhere they specify a `certificateAuthorityData`.
+Whenever `certificateAuthorityData` is empty and `secretName` is populated, Pinniped will expect a secret in the same
+namespace with that name.
+The secret must be of type `kubernetes.io/tls`.
+If the secret contains a key with name `ca.crt`, that value will be used as the CA bundle.
+If the secret does not contain a key with name `ca.crt`, but does have a key with name `tls.crt`, that value will be
+used as the CA bundle.
+Any other keys in the secret are ignored.
+
+Kubernetes itself defines a secret with type `kubernetes.io/tls` with keys `tls.crt` and `tls.key`.
+See [docs](https://kubernetes.io/docs/concepts/configuration/secret/#tls-secrets).
+
+This secret type is extended by `cert-manager` to include an optional key `ca.crt` which contains the ["most" root](https://cert-manager.io/docs/releases/release-notes/release-notes-1.4/#ca-vault-and-venafi-issuer-handling-of-cacrt-and-tlscrt)
+CA that `cert-manager` is aware of.
+The `cert-manager` documentation gives [this description](https://cert-manager.io/docs/trust/trust-manager/#cert-manager-integration-cacrt-vs-tlscrt)
+of when to use `ca.crt`  or `tls.crt`.
+Since `ca.crt` is the preference for `cert-manager`, Pinniped will prefer `ca.crt` when available.
+
+The existing `certificateAuthorityData` field can be populated with a certificate bundle, which eases rotation of
+certificates, especially across different PKI trees.
+Since `ca.crt` and `tls.crt` fields in a `kubernetes.io/tls` secret generally hold only a single certificate, Pinniped
+users may need to use `certificateAuthorityData` to accomplish this rotation without service interruption.
+
+### Validations and Status
+
+Since `certificateAuthorityData` is currently optional in all applicable custom resources, no secret is required.
+
+Most of these resources have a status condition of type `TLSConfigurationValid` (or something similar) which will be
+enhanced with any validations for either `certificateAuthorityData` or the given secret.
+
+It is a configuration error to specify both a `certificateAuthorityData` and a `secretName`, and the CR's status
+conditions will indicate this.
+
+When a `secretName` is specified, the secret must exist, have type `kubernetes.io/tls`, and have at least one
```

Comment:
Does cert-manager output Secrets as type `kubernetes.io/tls`?

Reply:
Also, what is the format of the data that cert-manager outputs into the keys in that Secret? Are we sure we know how to parse it?

Asking these questions because obviously we would want to be able to read Secrets from cert-manager.

Reply:
I couldn't find this explicitly documented anywhere. Secrets with type `kubernetes.io/tls` are shown in the examples at https://cert-manager.io/docs/usage/certificate/ and when I followed the setup at https://pinniped.dev/posts/v0-25-0-external-cert-mgmt-impersonation-proxy/ the secrets are of type `kubernetes.io/tls`.

However, I just realized another problem. Kubernetes requires key `tls.key` upon creation of a secret of type `kubernetes.io/tls`. This key would have no purpose and in fact would be dangerous to use a real key here.

Kubernetes rejects this secret:

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: secret-tls
type: kubernetes.io/tls
data:
  tls.crt: Zm9vCg== # must be base64, so let's use "echo foo | base64"
```

with error:
```
% kubectl apply -f <>
The Secret "secret-tls" is invalid: data[tls.key]: Required value
```

Reply:
It is possible to use any dummy value for `tls.key` (as long as it is base64-encoded) but that feels very confusing for users.

Reply:
Per our discussions, we will be accepting secrets (tls and opaque ) and configmaps as references to the CA data. Additionally, we will be allowing users to specify which key in the secret or configmap data pinniped will read to get the CA data. I've also added a section to the docs calling out security considerations that if a tls secret is being supplied, then pinniped will be able to read the key in the tls secret.

Marking this as resolved, but feel free to unresolved it if more discussions are required.